server/punishments.ts

```diff
@@ -923,6 +943,101 @@ export const Punishments = new class {
 		}
 	}
 
+	/**
+	 * Bans a user from using groupchats. Returns an array of roomids of the groupchat they created, if any.
+	 * We don't necessarily want to delete these, since we still need to warn the participants,
+	 * and make a modnote of the participant names, which doesn't seem appropriate for a Punishments method.
+	 */
+	groupchatBan(user: User | ID, expireTime: number | null, id: ID | null, reason: string | null) {
+		if (!expireTime) expireTime = Date.now() + GROUPCHATBAN_DURATION;
+		const punishment = ['GROUPCHATBAN', id, expireTime, reason] as Punishment;
+
+		const groupchatsCreated = [];
+		const targetUser = Users.get(user);
+		if (targetUser) {
+			for (const roomid of targetUser.inRooms || []) {
+				const targetRoom = Rooms.get(roomid);
+				if (!targetRoom?.roomid.startsWith('groupchat-')) continue;
+				if (targetRoom.game && targetRoom.game.removeBannedUser) {
+					targetRoom.game.removeBannedUser(targetUser);
+				}
+				targetUser.leaveRoom(targetRoom.roomid);
+
+				// Handle groupchats that the user created
+				if (targetRoom.auth.get(targetUser) === Users.HOST_SYMBOL) {
+					groupchatsCreated.push(targetRoom.roomid);
+					Punishments.bannedGroupchatParticipants[targetRoom.roomid] = new Set(
+						// Room#users is a UserTable where the keys are IDs,
+						// but typed as strings so that they can be used as object keys.
+						Object.keys(targetRoom.users) as ID[]
+					);
+				}
+			}
+		}
+
+		Punishments.roomPunish("groupchat", user, punishment);
```

Comment:
Yes, I already scoped it to a fake room.

Reply:
Oh wow, nice.

Reply:
I should probably have finished reading before posting preliminary thoughts.